lib/ansible/plugins/connection/ssh.py

```diff
@@ -309,8 +309,49 @@ def _examine_output(self, source, state, b_chunk, sudoable):
 
         return b''.join(output), remainder
 
+
     def _run(self, cmd, in_data, sudoable=True):
         '''
+        Will retry execution if ssh were unable to connect or timeout
+        '''
+        remaining_tries = int(C.ANSIBLE_SSH_RETRIES) + 1
+        cmd_summary = ' '.join(cmd)
+        for attempt in range(remaining_tries):
+            try:
+                return_tuple = self._run_once(cmd=cmd, in_data=in_data, sudoable=sudoable)
+                # 0 = success
+                # 1-254 = remote command return code
+                # 255 = failure from the ssh command itself
+                if return_tuple[0] != 255:
+                    break
+                else:
+                    raise AnsibleConnectionFailure("Failed to connect to the host via ssh.")
+            except (AnsibleConnectionFailure, AnsibleConnectionTimeout, Exception) as e:
+                if attempt == remaining_tries - 1:
+                    raise
+                else:
+                    pause = 2 ** attempt - 1
+                    if pause > 30:
+                        pause = 30
+
+                    if isinstance(e, AnsibleConnectionFailure):
+                        msg = "ssh_retry [%s]: attempt: %d, ssh return code is 255." % (self.host, attempt)
+                    elif isinstance(e, AnsibleConnectionTimeout):
+                        msg = "ssh_retry [%s]: attempt: %d, ssh timeout." % (self.host, attempt)
+                    else:
+                        msg = "ssh_retry [%s]: attempt: %d, caught exception(%s) from" % (self.host, attempt, e)
```

Comment:
Seems like this clause should be sufficient, especially if the AnsibleConnectionFailure has a reasonable message.
(and or, AnsibleConnectionTimeout, assuming it was meant to be raised somewhere...)